clintrials/visualization/models.py

```
@@ -68,3 +64,92 @@ def fmt(v):
+            for c in self.df.columns:
+                if c in cat_cols or self.df[c].nunique() < len(self.df) / 2:
+                    if len(grouping_cols) < 3:
+                        grouping_cols.append(c)
+            # Ensure we don't group by all columns
+            if len(grouping_cols) == len(self.df.columns):
+                grouping_cols = grouping_cols[:-1]
+
+        # Limit to 3 levels of nesting (4 levels total including leaf table)
+        grouping_cols = grouping_cols[:3]
+
+        def fmt(v):
```

Ruff (0.15.21)

[warning] Missing return type annotation for private function `fmt`

(ANN202)